modules/javafx.web/src/main/java/com/sun/webkit/network/PublicSuffixes.java

```diff
@@ -71,101 +89,134 @@ private PublicSuffixes() {
+                logger.finest("rules: {0}", toLogString(rules));
             }
-        } else {
-            logger.warning("Resource not found: [{0}]",
-                    resourceName);
         }
 
-        result = result != null
-                ? Collections.unmodifiableMap(result)
-                : Collections.<String,Rule>emptyMap();
-        if (logger.isLoggable(Level.FINEST)) {
-            logger.finest("result: {0}", toLogString(result));
+        static Rules getRules(String domain) {
+            String tld = getTopLevelDomain(domain);
+            if (tld.isEmpty()) {
+                return null;
+            }
+            return rulesCache.computeIfAbsent(tld, k -> createRules(tld));
         }
-        return result;
-    }
 
-    /**
-     * Loads the public suffix list from a given reader.
-     */
-    private static Map<String,Rule> loadRules(BufferedReader reader)
-        throws IOException
-    {
-        Map<String,Rule> result = new LinkedHashMap<String, Rule>();
-        String line;
-        while ((line = reader.readLine()) != null) {
-            line = line.split("\\s+", 2)[0];
-            if (line.length() == 0) {
-                continue;
+        private static String getTopLevelDomain(String domain) {
+            domain = IDN.toUnicode(domain, IDN.ALLOW_UNASSIGNED);
+            int n = domain.lastIndexOf('.');
+            if (n == -1) {
+                return domain;
             }
-            if (line.startsWith("//")) {
-                continue;
+            return domain.substring(n + 1);
+        }
+
+        private static Rules createRules(String tld) {
+            try (InputStream pubSuffixStream = getPubSuffixStream()) {
+                if (pubSuffixStream == null) {
+                    return null;
+                }
+                ZipInputStream zis = new ZipInputStream(pubSuffixStream);
+                ZipEntry ze = zis.getNextEntry();
+                while (ze != null) {
+                    if (ze.getName().equals(tld)) {
+                        return new Rules(zis);
+                    } else {
+                        ze = zis.getNextEntry();
+                    }
+                }
+            } catch (IOException ex) {
+                logger.warning("Unexpected error", ex);
             }
-            Rule rule;
-            if (line.startsWith("!")) {
-                line = line.substring(1);
-                rule = Rule.EXCEPTION_RULE;
-            } else if (line.startsWith("*.")) {
-                line = line.substring(2);
-                rule = Rule.WILDCARD_RULE;
+            return null;
+        }
+
+        private static InputStream getPubSuffixStream() {
+            InputStream is = AccessController.doPrivileged(
+                (PrivilegedAction<InputStream>) () -> {
+                    String javaHome = System.getProperty("java.home");
+                    String resourceName = "lib/security/public_suffix_list.dat";
```

Comment:
Maybe have the path to the file in a static final field so you don't have to duplicate it in two places?